pep-0xxx.rst

```diff
@@ -0,0 +1,797 @@
+        operation cannot complete until more data is read from the network, or
+        until more data is available in the input buffer.
+        """
+
+
+Certificates
+~~~~~~~~~~~~
+
+This module would define an abstract X509 certificate class. This class would
+have almost no behaviour, as the goal of this module is not to provide all
+possible relevant cryptographic functionality that could be provided by X509
+certificates. Instead, all we need is the ability to signal the source of a
+certificate to a concrete implementation.
+
+For that reason, this certificate implementation defines only constructors. In
+essence, the certificate object in this module could be as abstract as a handle
+that can be used to locate a specific certificate.
+
+Concrete implementations may choose to provide alternative constructors, e.g.
+to load certificates from HSMs. If a common interface emerges for doing this,
+this module may be updated to provide a standard constructor for this use-case
+as well.
+
+    class Certificate(metaclass=ABCMeta):
+        @abstractclassmethod
+        def from_buffer(self, buffer: bytes) -> Certificate:
+            """
+            Creates a Certificate object from a byte buffer. This byte buffer
+            may be either PEM-encoded or DER-encoded. If the buffer is PEM
+            encoded it *must* begin with the standard PEM preamble (a series of
+            dashes followed by the ASCII bytes "BEGIN CERTIFICATE" and another
+            series of dashes). In the absence of that preamble, the
+            implementation may assume that the certificate is DER-encoded
+            instead.
+            """
+
+        @abstractclassmethod
+        def from_file(self, path: Union[pathlib.Path, AnyStr]) -> Certificate:
+            """
+            Creates a Certificate object from a file on disk. This method may
+            be a convenience method that wraps ``open`` and ``from_buffer``,
+            but some TLS implementations may be able to provide more-secure or
+            faster methods of loading certificates that do not involve Python
+            code.
+            """
+
+
+Private Keys
+~~~~~~~~~~~~
+
+This module would define an abstract private key class. Much like the
+Certificate class, this class has almost no behaviour in order to give as much
+freedom as possible to the concrete implementations to treat keys carefully.
+
+This class has all the caveats of the ``Certificate`` class.
+
+    class PrivateKey(metaclass=ABCMeta):
+        @abstractclassmethod
+        def from_buffer(self,
+                        buffer: bytes,
+                        password=None: Optional[Union[Callable[[], Union[AnyStr, bytearray]], AnyStr, bytearray]) -> Certificate:
+            """
+            Creates a PrivateKey object from a byte buffer. This byte buffer
+            may be either PEM-encoded or DER-encoded. If the buffer is PEM
+            encoded it *must* begin with the standard PEM preamble (a series of
```

Comment:
Good question. I'm not honestly sure. I'm not even honestly sure that we care that much.

The risk is that simply defining a function that takes a path/buffer and returns a list of Certificate and a PrivateKey forces our Certificate and PrivateKey objects to become concretized: that is, they need to be backed by actual X509 objects, because they can't just all point at the same file anymore. This hurts some possible optimisations (e.g. having Certificate objects actually just be a wrapper to the path to the cert for OpenSSL).

Ultimately, I think this "common case" is actually an OpenSSL special (I don't know that other backends support this, and some definitely don't e.g. SecureTransport), so I'm inclined to say that it doesn't belong in the generic implementation.